### PR TITLE
wasi: do not panic on extreme file timestamps in DescriptorStat

### PR DESCRIPTION
**Pull request by SebTardif**
## Summary

`DescriptorStat::new` converted host `SystemTime` values to WASI `Datetime` with `unwrap()`. When conversion fails (time outside the representable `i64` second range), that panics the host while building file stats.

Map conversion failure to `None` (missing timestamp) on the existing `Option<Datetime>` fields.

## Failure scenario

Host `Metadata` can surface `SystemTime` values that do not fit WASI `Datetime` (for example extreme far-past times representable as `SystemTime` but not as `i64` seconds). A subsequent guest or host `stat` path that builds `DescriptorStat` would panic instead of returning stats with missing timestamp fields.

Guest `set_times` is not the reproducer for the overflow itself: those APIs already take `Datetime` values within range. The panic is on host-side conversion of `SystemTime` into `Datetime`.

## Origin

The previous description incorrectly attributed introduction to the cap-primitives migration (#13872). That PR refactored the conversion site; the unwrap-style failure mode is not claimed as newly introduced there. This change is a small hardening of the conversion.

## Test

Unit test under `crates/wasi/src/filesystem.rs` (`mod test`) asserts that extreme `SystemTime` values fail `Datetime::try_from` and map to `None` without panicking.

Integration coverage of the surrounding set_times/stat stack already exists as `p1_fd_filestat_set` in `crates/test-programs` / `crates/wasi/tests`. A guest program cannot construct the overflowing `SystemTime` via the public WASI timestamp APIs, so the edge case stays as a unit test on the conversion.

```
cargo test -p wasmtime-wasi --lib test::datetime_try_from
```

### CODE REVIEW

File: crates/wasi/src/filesystem.rs [hunk 1]
```diff
@@ -310,18 +310,29 @@ impl DescriptorStat {
...
     /// count.
     fn new(meta: &Metadata, link_count: u64) -> Self {
-        fn datetime_from(t: std::time::SystemTime) -> Datetime {
-            // FIXME make this infallible or handle errors properly
-            Datetime::try_from(t).unwrap()
+        // Extreme SystemTime values can overflow Datetime's i64 seconds range.
+        // Treat those as missing timestamps instead of panicking the host
+        // when a guest stats a preopened path after set_times.
+        fn datetime_from(t: std::time::SystemTime) -> Option<Datetime> {
+            Datetime::try_from(t).ok()
```
Comment (alexcrichton) [inline crates/wasi/src/filesystem.rs]: Given how small this is now, could this get inlined directly?

File: crates/wasi/src/filesystem.rs [hunk 2]
```diff
@@ -1167,3 +1178,26 @@ impl WasiFilesystemCtxView<'_> {
         Ok(results)
     }
 }
+
+#[cfg(test)]
+mod datetime_from_tests {
```
Comment (alexcrichton) [inline crates/wasi/src/filesystem.rs]: Conventionally these tests are in `mod test` as opposed to something more specific. Additionally we prefer to lean more heavily on integration tests exercise far more of the stack and ensure more complete fixes. Would it be possible to to model this as a test in `crates/test-programs/src/bin/*.rs`?

File: crates/wasi/src/filesystem.rs [hunk 3]
```diff
@@ -1167,3 +1178,26 @@ impl WasiFilesystemCtxView<'_> {
...
+    use crate::clocks::Datetime;
+    use std::time::{Duration, SystemTime};
+
+    /// Extreme SystemTime values can fail `Datetime::try_from`. The old
+    /// `datetime_from` path used `.unwrap()`, which would panic the host
+    /// during `DescriptorStat` construction. Mapping with `.ok()` must
+    /// yield `None` instead.
```
Comment (alexcrichton) [inline crates/wasi/src/filesystem.rs]: Please go through the output of your AI and consider rewording/removing comments like this. Comments like this about the state of a previous version of the tree are basically irrelevant noise as soon as the patch lands. The test here is self-explanatory and needs no comments.

**Review by alexcrichton — COMMENTED** [review]
Thanks for the PR, but please be sure you review our [AI tool usage policy](https://github.com/bytecodealliance/governance/blob/main/AI_TOOL_POLICY.md). You own the code and you own the responsibility for being correct here, and for example the analysis of when this was introduced was wrong. The PR referred there was a refactor and did not introduce this.

File: crates/wasi/src/filesystem.rs (hunk 1 above)
Comment (SebTardif) [reply crates/wasi/src/filesystem.rs]: Inlined the `Datetime::try_from(...).ok()` mapping at the three call sites.

File: crates/wasi/src/filesystem.rs (hunk 2 above)
Comment (SebTardif) [reply crates/wasi/src/filesystem.rs]: Renamed to `mod test` and dropped the narrative comments.

On modeling this in `crates/test-programs/src/bin/*.rs`: guest `set_times` / filestat APIs already take WASI `Datetime` (`i64` seconds), so a guest cannot construct a `SystemTime` outside that range through the public ABI. The panic path is host `Metadata` -> `Datetime` conversion. Existing integration coverage for the surrounding stack is `p1_fd_filestat_set`. The remaining unit test only covers the overflow mapping that guests cannot reach.

File: crates/wasi/src/filesystem.rs (hunk 3 above)
Comment (SebTardif) [reply crates/wasi/src/filesystem.rs]: Removed. Left a comment-free unit test that only asserts the conversion edge.

**Comment by SebTardif** [comment]
@alexcrichton

> please be sure you review our AI tool usage policy... the analysis of when this was introduced was wrong. The PR referred there was a refactor and did not introduce this.

Thanks for the review. I re-read the [AI tool use policy](https://github.com/bytecodealliance/governance/blob/main/AI_TOOL_POLICY.md) and corrected the PR description: #13872 was a refactor of the conversion site, not the introduction of this failure mode. Sorry for the incorrect origin claim.

Addressed the inline notes in the latest commit: inlined the conversion, `mod test` without narrative comments, and explained why a guest test-program cannot reproduce the host `SystemTime` overflow (WASI timestamps are already `Datetime`). Happy to adjust further if you want the unit test dropped entirely and only rely on existing `p1_fd_filestat_set` coverage.

File: crates/wasi/src/filesystem.rs (hunk 2 above)
Comment (alexcrichton) [reply crates/wasi/src/filesystem.rs]: Yes this can't be driven by purely guest-defined behavior, but it should be possible to prepare a filesystem on the host where there's a file with an extremely old timestamp in it which the guest attempts to stat and read. This can then assert that there's no panic and additionally the behavior of the timestamp basically just disappearing.

The test here is additionally overly-verbose and AI-generated where it checks `is_err()` and `ok().is_none()`. Please review all code generated by an AI, including tests.

Comment (SebTardif) [reply crates/wasi/src/filesystem.rs]: Done. Host prepares a file with an extreme mtime before the guest runs; the guest does `path_filestat_get` and reads the contents (`p1_stat_extreme_host_mtime` under `crates/test-programs`, wired through the p1 and p2 runners).

On macOS the OS clamps that extreme time into a far-past value that still fits `i64` seconds, so `Datetime::try_from` succeeds. The remaining failure mode was converting that negative time into wasi:clocks/wall-clock (`u64` seconds), which used to make the whole stat return `OVERFLOW`. Unrepresentable wall-clock times are now treated as missing timestamps so filestat succeeds and the file remains readable.

File: crates/wasi/tests/all/p1.rs [hunk 4]
```diff
@@ -25,6 +25,29 @@ async fn run(path: &str, with_builder: impl FnOnce(&mut WasiCtxBuilder)) -> Resu
     Ok(())
 }
 
+async fn run_with_workspace_setup(
```
Comment (alexcrichton) [inline crates/wasi/tests/all/p1.rs:13]: Can you redefine the above function in terms of this one? Otherwise this looks like duplicated functionality.

File: crates/wasi/tests/all/p2/sync.rs [hunk 5]
```diff
@@ -31,6 +31,34 @@ fn run(path: &str, with_builder: impl Fn(&mut WasiCtxBuilder)) -> Result<()> {
     Ok(())
 }
 
+fn run_with_workspace_setup(
```
Comment (alexcrichton) [inline crates/wasi/tests/all/p2/sync.rs:14]: Can the above function be implemented in terms of this one instead of duplicating it?

File: crates/wasi/tests/all/p2/async_.rs [hunk 6]
```diff
@@ -27,6 +27,30 @@ async fn run(path: &str, with_builder: impl FnOnce(&mut WasiCtxBuilder)) -> Resu
         .map_err(|()| wasmtime::format_err!("run returned a failure"))
 }
 
+async fn run_with_workspace_setup(
```
Comment (alexcrichton) [inline crates/wasi/tests/all/p2/async_.rs:14]: Can the above function be deduplicated with this one?

File: crates/wasi/src/filesystem.rs [hunk 7]
```diff
@@ -1167,3 +1171,54 @@ impl WasiFilesystemCtxView<'_> {
...
+            .read_to_end(&mut buf)
+            .unwrap();
+        assert_eq!(buf, b"hello");
+        let _ = (
+            stat.data_access_timestamp,
+            stat.data_modification_timestamp,
+            stat.status_change_timestamp,
+        );
+    }
+}
```
Comment (alexcrichton) [inline crates/wasi/src/filesystem.rs]: It's ok to drop this custom test module now that it's integration-tested with the additional in-wasm test

File: crates/wasi/tests/all/p1.rs [hunk 8]
```diff
@@ -69,6 +92,32 @@ async fn p1_fd_filestat_get() {
...
+            let path = dir.join("extreme.dat");
+            File::create(&path)?.write_all(b"hello")?;
+            let extreme = SystemTime::UNIX_EPOCH
+                .checked_sub(Duration::from_secs((i64::MAX as u64) + 1))
+                .expect("construct extreme SystemTime");
+            let f = File::options().write(true).open(&path)?;
+            let times = FileTimes::new().set_modified(extreme).set_accessed(extreme);
+            // Platforms may reject or clamp extreme times.
+            let _ = f.set_times(times);
+            Ok(())
```
Comment (alexcrichton) [inline crates/wasi/tests/all/p1.rs]: I might recommend extracting this to a helper shared function in `store.rs` to avoid the need to duplicate this across p1/p2-sync/p2-async

File: crates/wasi/tests/all/p1.rs (hunk 4 above)
Comment (SebTardif) [reply crates/wasi/tests/all/p1.rs:13]: @alexcrichton

> Can you redefine the above function in terms of this one?

Done. `run` is now `run_with_workspace_setup(path, |_| Ok(()), with_builder)`.

File: crates/wasi/tests/all/p2/sync.rs (hunk 5 above)
Comment (SebTardif) [reply crates/wasi/tests/all/p2/sync.rs:14]: @alexcrichton

> Can the above function be implemented in terms of this one?

Done for p2-sync as well: `run` delegates to `run_with_workspace_setup` with a no-op setup.

File: crates/wasi/tests/all/p2/async_.rs (hunk 6 above)
Comment (SebTardif) [reply crates/wasi/tests/all/p2/async_.rs:14]: @alexcrichton

> Can the above function be deduplicated with this one?

Done for p2-async the same way.

File: crates/wasi/src/filesystem.rs (hunk 7 above)
Comment (SebTardif) [reply crates/wasi/src/filesystem.rs]: @alexcrichton

> It's ok to drop this custom test module now that it's integration-tested

Removed the `#[cfg(test)] mod test` block from `filesystem.rs`.

File: crates/wasi/tests/all/p1.rs (hunk 8 above)
Comment (SebTardif) [reply crates/wasi/tests/all/p1.rs]: @alexcrichton

> extract this to a helper shared function in `store.rs`

Added `store::prepare_extreme_mtime_fixture` and wired p1 / p2-sync / p2-async `p1_stat_extreme_host_mtime` through it.

**Review by alexcrichton — APPROVED** [review]

**Comment by SebTardif** [comment]
Pushed a MinGW fix for the merge-queue failure on `Test MinGW x86_64 (wasmtime-wasi)`.

The extreme-mtime fixture used `SystemTime::UNIX_EPOCH - (i64::MAX+1)`, which is constructible on Unix but not on Windows FILETIME-based `SystemTime`, so setup panicked before the guest ran. The fixture now prefers that far-past value when available and falls back to the Windows epoch (~1601) otherwise. The guest still only asserts that `path_filestat_get` does not panic the host.

Happy to be re-queued when convenient.
